src/lib.rs

```diff
@@ -712,38 +714,43 @@ fn postprocess_chunks(
             !invalid
         });
     }
+}
 
-    // Find fdAT chunks and attempt to recompress them
-    // Note if there are multiple fdATs per frame then decompression will fail and nothing will change
-    let mut fdat: Vec<_> = png
-        .aux_chunks
-        .iter_mut()
-        .filter(|c| &c.name == b"fdAT")
-        .collect();
-    if opts.idat_recoding && !fdat.is_empty() {
-        let buffer_size = orig_ihdr.raw_data_size();
-        fdat.par_iter_mut()
-            .with_max_len(1)
-            .enumerate()
-            .for_each(|(i, c)| {
-                if deadline.passed() || c.data.len() <= 4 {
-                    return;
-                }
-                if let Ok(mut data) = deflate::inflate(&c.data[4..], buffer_size).and_then(|data| {
-                    let max_size = AtomicMin::new(Some(c.data.len() - 5));
-                    opts.deflate.deflate(&data, &max_size)
-                }) {
-                    debug!(
-                        "Recompressed fdAT #{:<2}: {} ({} bytes decrease)",
-                        i,
-                        c.data.len(),
-                        c.data.len() - 4 - data.len()
-                    );
-                    c.data.truncate(4);
-                    c.data.append(&mut data);
-                }
-            });
+/// Recompress the additional frames of an APNG
+fn recompress_frames(png: &mut PngData, opts: &Options, deadline: Arc<Deadline>) {
+    if !opts.idat_recoding || png.frames.is_empty() {
+        return;
     }
+    // Use the same filter chosen for the main image
+    // No filter means we failed to optimise the main image and we shouldn't bother trying here
+    let Some(filter) = png.filter else {
+        return;
+    };
```

Comment:
So I'd be interested in opinions on this. As the comment says, we're unlikely to be able to optimise the fdATs if we failed to optimise the IDAT. The exception is if the image had been previously optimised with the current version of oxipng (before this PR): We will fail to optimise the IDAT any further, but could still optimise the fdATs. The problem is that since IDAT optimisation failed, we won't know what filter to use for the fdATs.
There is a workaround though: using the `--force` option will force the IDAT optimisation, thus providing the necessary filter for the fdATs.

Reply:
Using `--force` to handle that case sounds like a reasonable first approach to me :+1: 

In theory, I guess there's nothing stopping individual `fdAT` frames from being better compressed using a filter different to that of the main `IDAT` image, especially for APNGs with animations featuring abrupt scene changes. However, performing such an exhaustive filter search is, in my opinion, better suited for an additional option or the "brute" mode, as it's likely to be quite slow.